src/storage/src/source/mod.rs

```diff
@@ -1003,37 +1056,182 @@ where
+                        Some(_) | None => continue 'partition_loop,
+                    };
+                    let (timestamp, _diff) = bindings.pop().unwrap();
+                    to_write
+                        .entry((timestamp, partition.clone()))
+                        .or_insert(Vec::new())
+                        .extend(data_messages.drain(..diff_usize));
                 }
             }
 
-            // Downgrade capability (if possible) before exiting
-            timestamp_histories.downgrade(cap, &partition_cursors);
-            source_metrics.capability.set(*cap.time());
-            // Downgrade compaction frontier to track the current time.
-            timestamp_histories.set_compaction_frontier(Antichain::from_elem(*cap.time()).borrow());
+            // Transmit the data for each durable timestamp.
+            for ((timestamp, _partition), messages) in to_write {
+                let mut msg_output = vec![];
+                for message in messages {
+                    // Update the limits of offsets we have transmitted, for capability downgrading.
+                    partition_cursors.insert(message.partition.clone(), message.offset + 1);
+                    // Form and enqueue the output.
+                    msg_output.push(Ok(SourceOutput::new(
+                        message.key,
+                        message.value,
+                        message.offset.offset,
+                        message.upstream_time_millis,
+                        message.partition,
+                        message.headers,
+                    )));
+                }
+                let ts_cap = cap.delayed(&timestamp);
+                output.session(&ts_cap).give_vec(&mut msg_output);
+            }
+
+            // We should now downgrade our capability.
+            // We must retain the ability to send any future data, which is determined by looking at
+            // the *next* offsets for each part we manage, as this indicates data we might still receive.
+            // XXX: note +1 on read_since
+            let r = cap.try_downgrade(read_since.elements().first().unwrap());
+            eprintln!("DOWNGRADE CAP RESULT: {:?}: {:?}", read_since, r);
+
+            // Mint new timestamps
+            let new_ts = cap
+                .try_delayed(&now())
+                .map(|c| *c.time())
+                .unwrap_or(*cap.time());
+            let new_ac = Antichain::from_elem(new_ts + 1);
+            let append_updates: Vec<_> = messages_by_part
+                .iter()
+                .filter_map(|(partition, messages)| {
+                    let diff: Diff = messages.len().try_into().unwrap();
+                    if diff == 0 {
+                        None
+                    } else {
+                        Some(((&(), partition), new_ts, diff))
+                    }
+                })
+                .collect();
+            let cas_result = futures::executor::block_on(write_handle.compare_and_append(
```

Comment:
This will not work with multiple workers. As soon as one worker mints bindings up to `read_since` that "blocks" other workers from minting bindings at `read_since`

Reply:
yes, this is on purpose! An invariant of this implementation is that only one reader should mint a given timestamp.  

For the "losing" worker, `cas_result` will be `Err(_)` and we'll immediately reactivate this worker so it can read in the minted timestamps and process messages.

Will that approach not work with the persist interface?

Reply:
What I was referring to is this: say we have two workers, `w1` and `w2`. Worker `w1` reads from partitions `p1` and `p2`, worker `w2` reads from partitions `w3`, and `w4`.

Now, worker `w1` mints bindings `((p1, 10), t, 1)` and `((p2, 10), t, 1)` and advances the frontier to `t+1`. This means worker `w2` can't "sneak in" any bindings for `p3` and `p4` at time `t`.

It's probably easier if we hop on a call to sort this out.

Reply:
I understand what you're saying now!

Reply:
Is it right to say in the current code that multiple workers will contend on this, and .. potentially some workers may starve if they are not able to commit their offsets?

Reply:
>  in the current code that multiple workers will contend on this
that's correct
> potentially some workers may starve if they are not able to commit their offsets
Also correct.  If they lose the `compare_and_append`, they'll be rescheduled and then attempt to commit timestamps for any (partition, offset) pair that still need timestamps.  Losing the `compare_and_append` results in immediate reactivation while winning the `compare_and_append` results in a delayed activation which mitigates that possibility somewhat -- but it is theoretically possible, yeah.  That's part of why I think moving to a "one special task/worker mints timestamps and the rest read" model is likely an improvement (but for the sake of getting something working in, I'd like to do as a follow-up)